README.md
# 微吉风

> 今天再大的事，到了明天就是小事；今年再大的事，到了明年就是故事。

实际效果可在微信中搜索同名小程序体验。


## 功能

- [x] 课表
  - [x] 背景图
  - [x] 日期
  - [x] 自定义
  - [ ] 无课表
  - [ ] 云查询
- [x] 校园卡
    - [x] 余额
    - [x] 消费记录
- [x] 宿舍电费
    - [x] 余量
    - [x] 用电记录
    - [x] 购电记录
    - [ ] 余量提醒
- [x] 教务成绩
- [x] 教学日历（外链版）
- [x] 图书馆
    - [x] 书目检索
    - [ ] 个人中心
- [x] 公告
- [ ] 体验帐号
- [ ] 校园运动圈（？）
- [ ] 电话本
- [ ] 留言板（？）
- [ ] 接入统一认证

什么时候会弃坑呢？


## 使用

将 `config-sample.js` 重命名为 `config.js`，即可在微信开发者工具中打开。

部分可能用得到的文档参见 [Wiki](https://github.com/choyri/WeGifun/wiki)。


## 开源类库

位于 [/utils/libs/](https://github.com/choyri/WeGifun/tree/master/utils/libs)。

- [countup.js](https://github.com/wux-weapp/wux-weapp/blob/master/src/countup/index.js)
- [es6-promise.js](https://github.com/stefanpenner/es6-promise)
- [object-path.js](https://github.com/mariocasciaro/object-path)
- [regenerator-runtime.js](https://github.com/facebook/regenerator/blob/master/packages/regenerator-runtime/runtime.js)
- [text-encoder-lite.js](https://github.com/coolaj86/TextEncoderLite/blob/master/text-encoder-lite.js)


## 交流

欢迎加入 QQ 群：`window.atob('NjM2NTkzNzA1')`。


## 赞赏

愿意请我喝奶茶吗？ ：）

<image width="200" src="https://img.choyri.com/reward.png">


## 许可

[WTFPL](https://zh.wikipedia.org/wiki/WTFPL)


CHANGELOG.md
### v1.3.5 (2019-06-22)

- **改善** 学期时间的计算

### v1.3.4 (2019-04-04)

- **修复** 连上课程的判断
- **改善** 小程序升级后强制刷新课表
- **改善** 对不支持查询电费余量的宿舍的友好提示

### v1.3.3 (2019-03-27)

- **新增** 隐藏自定义课程
- **修复** 更新课表后不显示自定义课表
- **修复** 同时间段连上课程的判断
- **修复** 连上课程被其他课程遮住
- **改善** 添加自定义课程时可直接设置持续节数
- **改善** 支持 Face ID
- **改善** 页面样式适配 iPhone 全面屏设备
- **改善** iOS 设备上下拉页面时的窗口背景色
- **改善** 书目检索封面懒加载
- **改善** 移动赞赏入口
- **改善** 细节调整

### v1.3.2 (2019-01-06)

- **修复** 继教同学无法获取课表
- **改善** 尚无课表时的提示

### v1.3.1 (2019-01-05)

- **新增** 点亮课表
- **修复** 课表头栏宽不一致

### v1.3.0 (2018-11-03)

- **新增** 自定义课表
- **改善** 课表的提示与指南
- **改善** 降低免上课程的透明度
- **改善** 开发模式下不上报错误

### v1.2.2 (2018-10-12)

- **修复** 书目检索图书信息页链接地址复制无效
- **修复** 生物认证接口在 FaceID 设备上的兼容
- **修复** 生物认证接口出错时没有正确回退到密码认证
- **改善** 生物认证接口的检测

### v1.2.1 (2018-10-06)

- **修复** 学号长度限制放宽
- **改善** 绑定文案
- **改善** 错误上报的内容

### v1.2.0 (2018-10-05)

- **新增** 语言设置
- **新增** 教学日历（外链版）
- **修复** 无课表数据时渲染课表错误
- **修复** 获取不到页面元素高度时报错
- **改善** 学号长度限制放宽
- **改善** 错误上报的步骤和内容
- **改善** 读取本地数据时不覆盖预设数据

### v1.1.4 (2018-09-22)

- **修复** 不存在指定宿舍时多余的错误弹窗
- **修复** 某些情况下渲染课表的错误
- **改善** 过滤不必要的错误
- **改善** 宿舍电费可用楼栋

### v1.1.3 (2018-09-09)

- **修复** 无法查看第六大节课程的详情

### v1.1.2 (2018-09-09)

- **新增** 错误警报机制
- **修复** 周数为0时的错误判断

### v1.1.1 (2018-09-06)

- **修复** 处于第0周时无法获取课表
- **修复** 非期末时期获取课表页的学期值错误

### v1.1.0 (2018-08-20)

- **新增** CET 准考证
- **修复** 兼容教师帐户获取学年学期
- **修复** 无网络状态下获取课表时显示两次错误弹窗
- **改善** 细节优化

### v1.0.3 (2018-07-11)

- **改善** 支持绑定教师帐户
- **改善** 未绑定状态下可以使用宿舍电费的部分模块
- **改善** 加快提示框的显示

### v1.0.2 (2018-07-08)

- **新增** 微信登录
- **新增** 可以选择隐藏非本周课程
- **改善** 宿舍设置提示
- **改善** 调整课程块背景颜色
- **改善** 取消体育课的独立课程时间表
- **改善** 更改赞赏位置

### v1.0.1 (2018-06-30)

- **新增** 支持获取指定学期的课表
- **新增** 公告模块
- **改善** 课程成绩各科目结果增加绩点项
- **改善** 细节优化

### v1.0.0 (2018-06-24)

- 重构与优化
- **新增** 课表页显示日期
- **新增** 分享

### v0.6.7 (2018-04-22)

- **新增** 电费充值支持指纹验证身份

### v0.6.6 (2018-04-21)

- **改善** 调整用户微信资料的获取和授权操作

### v0.6.5 (2018-04-08)

- **改善** 调整文案

### v0.6.4 (2018-04-06)

- **修复** 课表为空时切换周数导致的错误
- **改善** 调整引导文案
- **改善** 获取用户信息时不携带登录态信息

### v0.6.3 (2018-04-01)

- **新增** 课表周数切换

### v0.6.2 (2018-03-28)

- **修复** iOS 不显示课表背景
- **改善** 调整 iOS 上下拖动窗口时的背景颜色

### v0.6.1 (2018-03-27)

- **新增** 课表背景设置
- **新增** 新版特性提示

### v0.6.0 (2018-03-26)

- **新增** 书目查询

### v0.5.1 (2018-03-07)

- **改善** 显示课程时间
- **改善** 调整课程块背景颜色

### v0.5.0 (2017-12-14)

- **新增** 课程成绩查询

### v0.4.5 (2017-11-19)

- **修复** 用户微信资料不能及时更新
- **改善** 将「服务」页入口从点击头像改为独立的跳转按钮

### v0.4.4 (2017-11-17)

- **改善** 不再自动跳转到「我的」
- **新增** 使用引导页

### v0.4.3 (2017-11-12)

- **修复** 属性「scope.userInfo」未定义
- **修复** 「event」的一些未定义错误
- **改善** 不选择房间的话无法使用电费服务

### v0.4.2 (2017-10-25)

- **修复** 跟在连上课程后的普通课程受其影响而隐藏
- **修复** 没有注销掉之前订阅的事件
- **修复** 电费记录页面一个不知道是什么也不知道是如何产生的Bug
- **更新** 支持第11-12小节课程
- **更新** 新的课程块背景颜色
- **更新** 第一次使用会有绑定服务的提示

### v0.4.1 (2017-10-18)

- **修复** 购电成功时消息提示框一闪而过
- **修复** 判断网络请求结果时的逻辑错误

### v0.4.0 (2017-10-16)

- **修复** 第9-11小节课程块样式错误
- **修复** 未绑定教务系统但显示了设置入口
- **更新** 优化同一时间课程的显示方案
- **更新** 优化网络请求的处理方案
- **更新** 优化了部分样式和逻辑
- **新增** 宿舍电费

### v0.3.1 (2017-09-04)

- **修复** 「周几」标题错误
- **修复** 没课的时候不显示「周几」
- **修复** 用户状态更改时界面刷新出错
- **更新** Tab栏改用原生上边框

### v0.3.0 (2017-09-03)

- **更新** 修改获取用户信息的方案
- **更新** 修改课表解析与渲染的方案
- **更新** 取消隐藏因单双周而不用上的课程
- **更新** 反馈途径更改为客服会话
- **更新** 修改课表的存储格式
- **更新** 优化了外观样式
- **更新** 重构了程序

### v0.2.2 (2017-07-14)

- **更新** 改进对连上课程的处理
- **更新** 不再支持特殊课程

### v0.2.1 (2017-07-07)

- **修复** 未在教学周期内时课程不显示
- **修复** 存在连上的课程时导致课表无法显示

### v0.2.0 (2017-06-11)

- **更新** 优化了外观样式
- **更新** 重构了程序
- **新增** 校园卡服务
- **新增** 获取用户微信资料，显示头像和昵称
- **新增** 多语言，当前支持简体中文和英文

### v0.1.0 (2017-04-01)

- 项目发布

services/edu.js
import regeneratorRuntime from '../utils/libs/regenerator-runtime'

// 调色板 # 课程块背景颜色 每门课一种 当前内置十二种
const SCHEDULE_PALETTE = [
  '#ff8a98', '#29b6f6', '#26a69a', '#7986cb', '#ff7043', '#4caf50',
  '#af5e53', '#48829e', '#ba68c8', '#eb586f', '#ffa726', '#c75080',
]

// 课程时间表
const TIME_TABLE = [
    ['08:20', '09:05'],
    ['09:15', '10:00'],
    [
      ['10:20', '10:30'],
      ['11:05', '11:15'],
    ],
    [
      ['11:15', '11:25'],
      ['12:00', '12:10'],
    ],
    ['14:00', '14:45'],
    ['14:55', '15:40'],
    ['16:00', '16:45'],
    ['16:55', '17:40'],
    ['18:40', '19:25'],
    ['19:35', '20:20'],
    ['20:30', '21:15'],
    ['21:25', '22:10'],
  ]

class Edu {
  static _currWeek = null

  static async getCurrWeek (refresh = false) {
    if (!refresh && this._currWeek) {
      return this._currWeek
    }

    let startDate = this._getSchoolStartDate()

    if (refresh || !startDate) {
      startDate = await wx.ooRequest.getSchoolStartDate()
      if (!startDate) {
        return
      }
      this._saveSchoolStartDate(startDate)
    }

    // 7 * 24 * 60 * 60 * 1000 = 604800000
    return this._currWeek = Math.ceil(((new Date()).getTime() - (new Date(startDate)).getTime()) / 604800000)
  }

  static async fetchSchedule (schoolTime) {
    schoolTime = schoolTime || this.getSchoolTime()

    const rawSchedule = await wx.ooRequest.getEduSchedule(schoolTime)
    if (!rawSchedule) {
      return
    }

    wx.ooSaveData({ edu: { schedule: rawSchedule } })

    this.saveSchedule(this.renderSchedule())

    return rawSchedule
  }

  static getSchedule () {
    return wx.ooCache.schedule && wx.ooCache.schedule.data || null
  }

  static renderSchedule (schedule, currWeek) {
    schedule = schedule || this._getSchoolSchedule()

    if (!schedule) {
      return null
    }

    schedule = wx.ooUtil.copy(schedule)
    currWeek = parseInt(currWeek || this._currWeek)

    let res = [],
      scheduleBg = {},
      scheduleBgIndex = Math.floor(Math.random() * 12)

    let dayIndex,
      sectionIndex

    const _pushCourse = course => {
      res[dayIndex] = res[dayIndex] || {
        unique: `week${currWeek}_day${course.day}`,
        index: dayIndex,
        data: [],
      }

      res[dayIndex].data[sectionIndex] = res[dayIndex].data[sectionIndex] || {
        unique: `week${currWeek}_day${course.day}_section${course.section}`,
        index: sectionIndex,
        data: [],
      }

      res[dayIndex].data[sectionIndex].data.push({
        data: course,
        index: res[dayIndex].data[sectionIndex].data.length,
        unique: `week${currWeek}_day${course.day}_section${course.section}_course${res[dayIndex].data[sectionIndex].data.length}`,
      })
    }

    for (let course of schedule) {
      // 课程数据的排序从一开始 减一以配合数组
      dayIndex = course.day - 1
      sectionIndex = course.section - 1

      // 是否需要上课
      course.attend = true

      // 课程是否可视 # 同一时间有多节课的时候隐藏不相干的课
      course.display = true

      // 课程周期范围
      course.weekRange = course.week.split(':').map(item => parseInt(item))

      const outOfRange = currWeek < course.weekRange[0] || currWeek > course.weekRange[1] || false

      // 单周时双周的课 & 双周时单周的课 & 不在周期内的课 不需要上
      if ((course.weekRange[2] === 1 && currWeek % 2 === 0) || (course.weekRange[2] === 2 && currWeek % 2 === 1) || outOfRange) {
        course.attend = false
      }

      // 同一时间段不同周有不同的课 # e.g. 1-15周(单),2-16周(双) / 1-3周(单),4-16周
      const currSection = res[dayIndex] && res[dayIndex].data[sectionIndex] || null
      if (currSection && currSection.data.length > 0) {
        // 当前课程不在周期内的话 跳过后续处理
        if (outOfRange) {
          continue
        }

        let affectedBySiblingCourse = false

        for (let i = 0; i < currSection.data.length; i++) {
          const siblingCourse = currSection.data[i].data

          // 一般情况下 隔壁课程需要上 那就说明当前课程不用上 # 此时 course.attend 是 false
          // 不一般的情况下 特殊课程 不考虑
          if (siblingCourse.attend && (siblingCourse.display || siblingCourse.forwardFrom !== undefined)) {
            affectedBySiblingCourse = true
            continue
          }

          // 如果隔壁课程不用上那就隐藏它
          if (siblingCourse.attend === false) {
            res[dayIndex].data[sectionIndex].data[i].data.display = false
          }
        }

        // 如果已有需要上的隔壁课程 跳过当前课程
        if (affectedBySiblingCourse) {
          continue
        }
      }

      // 连上课程
      const prevSectionIndex = sectionIndex - 1,
        prevSection = res[dayIndex] && res[dayIndex].data[prevSectionIndex] || null

      let affectedByPrevSection = false

      // 如果当前课程偏移 那么即使前一节有课也不管
      if (prevSection && course.offset != 2) {
        const currValue = course.name + course.teacher + course.location + course.attend

        for (const _prevCourse of prevSection.data) {
          const prevCourse = _prevCourse.data,
            prevValue = prevCourse.name + prevCourse.teacher + prevCourse.location + prevCourse.attend

          // 如果前一节课偏移就不要
          if (prevCourse.offset == 1) {
            continue
          }

          // 如果前一节课和本节课不同 跳过
          if (prevValue !== currValue) {
            continue
          }

          // 如果两节课的周期不同 跳过
          if (prevCourse.weekRange[1] < course.weekRange[0] || prevCourse.weekRange[0] > course.weekRange[0]) {
            continue
          }

          // 此时 前一节课和本节课相同 连上课程

          // 隐藏当前课程
          course.display = false

          // 给当前课程打标记 第几节开始的课 # 因为值可能是数字零 所以此处不能用 || 的写法
          course.forwardFrom = prevCourse.forwardFrom !== undefined ? prevCourse.forwardFrom : prevSectionIndex

          // 修改连上课程的高度 # 因为前面那个时间段可能不止一门课 所以用循环找出目标
          const headSection = res[dayIndex].data[course.forwardFrom]
          for (const _headCourse of headSection.data) {
            const headCourse = _headCourse.data

            if (headCourse.name !== course.name || headCourse.attend !== course.attend) {
              continue
            }

            // 每个课程块高度 200  间隔 10 # 例如 两节连上的话高度就为 410

            // 偏移课程高度为 100 # 第六大节如果是连上课程 说明和第五大节连上 目前来看只有 9-11 的情况 所以高度也为 100
            const _offset = (headCourse.offset || course.offset || course.section == 6) ? 100 : 200

            res[dayIndex].data[course.forwardFrom].data[_headCourse.index].data
              .height = (sectionIndex - course.forwardFrom) * 210 + _offset

            break
          }

          affectedByPrevSection = true

          break
        }

        if (affectedByPrevSection) {
          // 后面不用判断 直接打入课程 进行下一轮循环
          _pushCourse(course)
          continue
        }
      }

      // 处理节次偏移
      if (course.offset) {
        course.top = course.offset == 1 ? 0 : 100
        course.height = 100
      }

      // 课程背景色 # 不用上的课没有背景色 即默认的灰色
      if (wx.ooCache.setting.lightUpSchedule || (outOfRange === false && course.attend)) {
        // 每门课一种颜色 # 以课程名字当索引
        const bgKey = course.name
        scheduleBg[bgKey] = scheduleBg[bgKey] || SCHEDULE_PALETTE[scheduleBgIndex++ % SCHEDULE_PALETTE.length]

        course.bg = scheduleBg[bgKey]
      }

      // 隐藏超长的连上课程 # e.g. 1-8节的课A 这周不用上 5-8节有课B 这时候隐藏课A
      for (let i = 0; i < sectionIndex; i++) {
        const sectionSets = res[dayIndex] && res[dayIndex].data[i] || null
        if (!sectionSets) {
          continue
        }

        for (let targetCourse of sectionSets.data) {
          if (targetCourse.data.height === undefined || targetCourse.data.forwardFrom !== undefined) {
            continue
          }

          if ((course.section - targetCourse.data.section) * 210 > targetCourse.data.height) {
            continue
          }

          if (targetCourse.data.display === true) {
            targetCourse.data.display = false
          }
        }
      }

      // 隐藏非本周课程
      if (course.display && !course.attend && currWeek > 0 && currWeek < 17 && wx.ooCache.setting.hideCourse) {
        continue
      }

      // 循环结束 打入课程
      _pushCourse(course)
    }

    // 周一到周五 即使某一天没课也给其赋空值
    for (let i = 0; i < 5; i++) {
      res[i] = res[i] || {}
    }

    return res
  }

  static saveSchedule (data) {
    wx.ooEvent.emit('updateSchedule', data)
    wx.ooSaveData({ schedule: { data, week: this._currWeek } })
  }

  static updateSchedule (force) {
    if (!this._getSchoolSchedule()) {
      console.log('课表数据不存在')
      return
    }

    if (this._currWeek === null) {
      console.log('当前周数不存在')
      return
    }

    if (this._currWeek === this._getScheduleWeek() && !force) {
      console.log('周数无变化')
      return
    }

    console.log(force ? '强制' : '周数改变', '重新渲染课表')

    this.saveSchedule(this.renderSchedule())
  }

  static getScheduleBg () {
    return wx.ooCache.schedule && wx.ooCache.schedule.bg || { path: '', style: 0 }
  }

  static saveScheduleBg (value) {
    let bg = this.getScheduleBg()
    Object.assign(bg, value)

    wx.ooEvent.emit('updateScheduleBg', bg)
    wx.ooSaveData({ schedule: { bg } })
  }

  static _getScheduleWeek () {
    return wx.ooCache.schedule ? wx.ooCache.schedule.week : 1
  }

  static _getSchoolStartDate () {
    return wx.ooCache.edu && wx.ooCache.edu.startDate || null
  }

  static _getSchoolSchedule () {
    const schedule = wx.ooCache.edu && wx.ooCache.edu.schedule || []
    const customSchedule = !wx.ooCache.setting.hideCustomCourse && wx.ooCache.edu && wx.ooCache.edu.schedule_custom || []

    return [...schedule, ...customSchedule]
  }

  static _saveSchoolStartDate (startDate) {
    wx.ooSaveData({ edu: { startDate } })
  }

  static getSchoolTime (forSchedule = false) {
    // 默认值
    let res = { grade: 5, semester: 2 }

    const _USER_ACCOUNT = wx.ooService.user.getAccount()

    if (_USER_ACCOUNT.id.length < 8) {
      return res
    }

    const FURTHER_EDUCATION = _USER_ACCOUNT.id.substr(0, 1) === 'Z'

    // 普通学号格式为 0118xxxx # 前两位院系 接下来两位入学年份
    // 继教学号格式为 Z18xxxxx # 只包含入学年份 无法判断院系
    const STUDENT_ID = !FURTHER_EDUCATION ? _USER_ACCOUNT.id : '_' + _USER_ACCOUNT.id

    const _DATE = new Date(),
      CURR_YEAR = _DATE.getFullYear(),
      CURR_MONTH = _DATE.getMonth() + 1,
      COLLEGE = STUDENT_ID.substr(0, 2),
      HIGHEST_GRADE = (FURTHER_EDUCATION || COLLEGE == '11') ? 5 : 4,
      ENROLLMENT_YEAR = 2000 + parseInt(STUDENT_ID.substr(2, 2))

    res.grade = CURR_YEAR - ENROLLMENT_YEAR

    if (CURR_MONTH > 8) {
      res.grade++
    }

    // 设置最高年级 # 建筑大五 其他大四
    if (res.grade > 4) {
      res.grade = HIGHEST_GRADE
    }

    // 学期
    res.semester = CURR_MONTH < 3 || CURR_MONTH > 8 ? 1 : 2

    res.gradeUpper = res.grade

    const isNormalMonth = () => (CURR_MONTH > 2 && CURR_MONTH < 6) || CURR_MONTH > 8,
      isLatestSemester = () => res.grade === HIGHEST_GRADE && res.semester === 2

    // 为课表获取学期时 指定情况下多增加一学期 # 非正常月份 即只有学期末才加 / 非最后学年学期
    if (forSchedule && !isNormalMonth() && !isLatestSemester()) {
      if (CURR_MONTH >= 6) {
        res.gradeUpper++
      }
    }

    return res
  }

  static async fetchScore (schoolTime) {
    return await wx.ooRequest.getEduScore(schoolTime)
  }

  static getTimeTable (course) {
    // 大节计数
    const sectionCount = (course.height === undefined || course.height == 100) ? 1 : Math.floor(course.height / 200)

    // 开始小节的下标
    const startSmallSection = course.section * 2 - 1

    // 大节组 # 储存课程下标
    let sectionGroup = []
    for (let i = 0; i < sectionCount; i++) {
      const index = startSmallSection + 2 * i
      sectionGroup.push({ start: index, end: index + 1 })
    }

    // 课程偏移 # sectionGroup 数组头部就是需要调整的值 所以直接指定下标为 0
    if (course.offset == 1) {
      // 偏上的就将结束下标减一
      sectionGroup[0].end--
    } else if (course.offset == 2) {
      // 偏下的就将开始下标加一
      sectionGroup[0].start++
    }

    // 带有小节的连上课程 类似于 3 小节 # e.g. 9-11
    if (course.height && course.height % 200 > 100) {
      // 最后一小节的下标
      const index = (course.section + sectionCount - 1) * 2 + 1

      // 如果是偏移课程 以偏移小节开始 普通大节结束 # 这种排课暂时未遇见 # e.g. 6-8
      let tmp = course.offset == 2 ? 1 : 0

      sectionGroup.push({ start: index, end: index + tmp })
    }

    // 是否较快下课
    const isFast = course.location.match(/二教|实验|室/g) === null

    let res = []

    for (const section of sectionGroup) {
      let startTime, endTime

      startTime = TIME_TABLE[section.start - 1][0]
      endTime = TIME_TABLE[section.end - 1][1]

      // 如果是第 3 4 节开始的课程 还需要再细分
      if (section.start === 3 || section.start === 4) {
        startTime = isFast ? startTime[0] : startTime[1]
        endTime = isFast ? endTime[0] : endTime[1]
      }

      res.push(`${startTime}-${endTime}`)
    }

    return res.join(' ')
  }

  static async fetchUserInfo () {
    return await wx.ooRequest.getEduUserInfo()
  }
}

export default Edu
